src/main/java/me/magnet/consultant/RoutingStrategies.java

```diff
@@ -0,0 +1,110 @@
+package me.magnet.consultant;
+
+import java.util.List;
+import java.util.Random;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Sets;
+
+public class RoutingStrategies {
+
+	public static final RoutingStrategy RANDOMIZED_WEIGHTED_DISTANCE = new RoutingStrategy() {
+
+		private final Random random = new Random();
+
+		@Override
+		public List<ServiceInstance> listInstances(ServiceLocator serviceLocator, String serviceName) {
+			List<ServiceInstance> instances = serviceLocator.listInstances(serviceName);
+			if (instances.size() <= 1) {
+				// No routing to do.
+				return instances;
+			}
+
+			// Pick a random number between 0-1.
+			double chance = random.nextDouble();
+
+			/*
+			 * Pick the index of the first instance to try based on this chance.
+			 * If the chance was < 0.5 then use the instance on index 0.
+			 * If the chance was >= 0.5 and < 0.75 then use the instance on index 1.
+			 * If the chance was >= 0.75 and < 0.875 then use the instance on index 2.
+			 * If the chance was >= 0.875 and < 0.9375 then use the instance on index 3.
+			 * Etc...
+			 */
+			int index = 0;
+			double threshold = 0.5;
+			while (threshold < chance && index < instances.size() - 1) {
+				index++;
+				threshold = Math.pow(0.5, index + 1);
+			}
+
+			return splice(instances, index);
+		}
+	};
+
+	public static final RoutingStrategy ROUND_ROBIN = new RoutingStrategy() {
+
+		private final Multimap<String, ServiceInstance> lastRequested = HashMultimap.create();
+
+		@Override
+		public List<ServiceInstance> listInstances(ServiceLocator serviceLocator, String serviceName) {
+			List<ServiceInstance> nextAvailable = serviceLocator.listInstances(serviceName);
+			Set<ServiceInstance> lastRequests = Sets.newHashSet(lastRequested.get(serviceName));
+
+			// If all instances have been hit at least once, reset the counter for this service.
+			if (lastRequests.containsAll(nextAvailable)) {
+				lastRequested.removeAll(serviceName);
+			}
+
+			// List the instances to hit in order of not hit to already hit (conserving network distance).
+			List<ServiceInstance> nextUp = nextAvailable.stream()
+					.filter(next -> !lastRequests.contains(next))
+					.collect(Collectors.toList());
+			nextAvailable.stream()
+					.filter(lastRequests::contains)
+					.forEachOrdered(nextUp::add);
+
+			// Register that nextUp[0] instance is going to get hit first.
+			if (!nextUp.isEmpty()) {
+				lastRequested.put(serviceName, nextUp.get(0));
+			}
+
+			return nextUp;
+		}
+	};
```

Comment:
We're talking in circles again:

- Single items *do* allow for proper bookkeeping, and *do* have solutions for distinctly different instances & racing (with the help of either a blacklist or transactions). 
- Lists and Iteraters can't do proper bookkeeping, we agreed already upon that above. Lists can't do any bookkeeping, and iterators are easily abused by converting to other structures (and they most likely will be accidentally abused when we use them in streams or RX). 
- sessions/transactions != list or iterator. Lists don't have any bookkeeping, and iterators are easily accidentally used incorrectly. I'm talking about a transaction that only exposes a `onNextInstance`/`getNextInstance` method, which *can* do all the bookkeeping. 

Because of the first point lists aren't necessary, and because of the others lists are the inferior solution, with significant usability problems.

> > It doesn't guarantee that, because of the concerns above
> It does, code here needs to be updated, but it does.

Lists fundamentally cannot guarantee that. Code example:

```java
List<ServiceInstance> instances = listInstances();
try {
    doSomethingWith(instances.get(0));
} catch (SomeException e) {
    //fallover
    doSomethingWith(instances.get(1)); //oops, we've invalidated the guarantee
}
```
`listInstances` doesn't register the use of `instances.get(1)`, and as such cannot possibly guarantee that each instance is used equally often. This pattern above is already present in our code base, and a very likely pattern to be used i.c.w. round robin.

So we can conclude that the round robin strategy:
- Doesn't round robin
- Doesn't guarantee any order
- Doesn't guarantee equal usage of instances

Which basically makes the round robin strategy useless.